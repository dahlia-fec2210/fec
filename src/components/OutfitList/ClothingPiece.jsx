/* eslint-disable react/jsx-no-bind */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable import/extensions */
/* eslint-disable react/prop-types */
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { RotatingLines } from 'react-loader-spinner';
import Price from '../relatedItems/Price.jsx';
import Star from '../common/Star.jsx';
import RemoveButton from './RemoveButton.jsx';
import OutfitCardPhoto from './OutfitCardPhoto.jsx';

const serverRoute = `http://localhost:${process.env.PORT}`;

function ClothingPiece({
  clothingPiece, outfit, setOutfit, left,
}) {
  const [price, setPrice] = useState(0);
  const [photo, setPhoto] = useState(null);
  const [salePrice, setSalePrice] = useState(0);
  const [productInfo, setProductInfo] = useState({});
  const [average, setAverage] = useState(0);

  const carouselStyle = {
    transform: `translate(-${left}px, 0)`,
    transition: 'transform 900ms ease-in-out',
  };

  useEffect(() => {
    const cache = JSON.parse(localStorage.getItem(`productInfo-${clothingPiece}`));
    if (cache === null) {
      axios.get(`${serverRoute}/products/${clothingPiece}`)
        .then((data) => {
          localStorage.setItem(`productInfo-${clothingPiece}`, data.data);
          setProductInfo(data.data);
        });
    } else {
      setProductInfo(cache);
    }
  }, []);

  useEffect(() => {
    const cache = JSON.parse(localStorage.getItem(`productData-${clothingPiece}`));
    if (cache === null) {
      axios.get(`${serverRoute}/products/${clothingPiece}/styles`)
        .then((response) => {
          cache.photos = response.data.results[0].photos[0].url;
          cache.price = response.data.results[0].original_price;
          cache.salePrice = response.data.results[0].sale_price;
          localStorage.setItem(`productData-${clothingPiece}`, JSON.stringify(cache));
          setPhoto(response.data.results[0].photos[0].url);
          setPrice(response.data.results[0].original_price);
          setSalePrice(response.data.results[0].sale_price);
        });
    } else {
      setPhoto(cache.photos[0]);
      setPrice(cache.price);
      setSalePrice(cache.salePrice);
    }
  }, [outfit]);

  useEffect(() => {
    const cache = JSON.parse(localStorage.getItem('averageReview')) || {};
    if (!cache[clothingPiece]) {
      axios.get(`${serverRoute}/reviews/meta/?product_id=${clothingPiece}`)
        .then((data) => {
          const reviews = data.data.ratings;
          const keys = Object.keys(reviews);
          let sum = 0;
          let numReviews = 0;
          keys.forEach((key) => {
            sum += (key * reviews[key]);
            numReviews += Number(reviews[key]);
          });
          cache[clothingPiece] = sum / numReviews;
          localStorage.setItem('averageReview', JSON.stringify(cache));
          setAverage(sum / numReviews);
        });
    } else {
      setAverage(cache[clothingPiece]);
    }
  }, []);

  function removeFromOutfit(event) {
    event.preventDefault();
    const newOutfit = outfit.filter((piece) => piece !== clothingPiece);
    setOutfit(newOutfit);
    localStorage.setItem('outfit', JSON.stringify(newOutfit));
    axios.post(`${serverRoute}/interactions`, { element: event.target.id, widget: 'Related Products', time: new Date().toTimeString() });
  }

  if (clothingPiece !== null && productInfo !== {}) {
    return (
      <div>
        <div className="related-product-card" style={carouselStyle}>
          <RemoveButton removeFromOutfit={removeFromOutfit} clothingPiece={clothingPiece} />
          <div>
            <OutfitCardPhoto photo={photo} clothingPiece={clothingPiece} />
            <div className="related-category">{productInfo.category}</div>
            <div className="related-name">{productInfo.name}</div>
            <Price
              price={price}
              salesPrices={salePrice}
            />
            <div className="related-stars">
              <Star percentage={(average / 5) * 100} />
            </div>
          </div>
        </div>
      </div>
    );
  }
  return (
    <RotatingLines
      strokeColor="grey"
      strokeWidth="5"
      animationDuration="0.75"
      width="96"
      visible
    />
  );
}

export default ClothingPiece;
