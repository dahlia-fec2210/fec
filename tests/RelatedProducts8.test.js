/* eslint-disable no-undef */
/* eslint-disable import/extensions */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import ClothingPiece from '../src/components/OutfitList/ClothingPiece.jsx';

test('renders snapshot for clothing piece', async () => {
  const tree = renderer
    .create(<ClothingPiece />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
