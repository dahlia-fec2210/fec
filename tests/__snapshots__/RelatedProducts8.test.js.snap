// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot for clothing piece 1`] = `
<div>
  <div
    className="related-product-card"
    style={
      {
        "transform": "translate(-undefinedpx, 0)",
        "transition": "transform 900ms ease-in-out",
      }
    }
  >
    <div
      className="related-stack"
      onClick={[Function]}
    >
      <div
        className="fa-stack"
        style={
          {
            "verticalAlign": "top",
          }
        }
      >
        <i
          className="related-circle fa-solid fa-regular fa-circle fa-stack-2x"
        />
        <i
          className="related-star fa-solid fa-x fa-stack-1x"
          id="outfit-list-remove-button:undefined"
        />
      </div>
    </div>
    <div>
      <img
        className="related-outfit-photo"
        src="https://img.ltwebstatic.com/images3_pi/2022/04/06/16492430704a5786a3329d6838490cfcc903aa6996_thumbnail_600x.webp"
      />
      <div
        className="related-category"
      />
      <div
        className="related-name"
      />
      <div
        className="related-price"
        data-testid="price"
      >
        $
        0
      </div>
      <div
        className="related-stars"
      >
        <div
          className="ratings"
        >
          <div
            className="empty-stars"
          >
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
          </div>
          <div
            className="full-stars"
            style={
              {
                "width": "0%",
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
